src/libs/globbing/globbing.c

```diff
@@ -0,0 +1,212 @@
+/**
+ * @file
+ *
+ * @brief Library for performing globbing on keynames.
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdb.h>
+#include <kdbglobbing.h>
+#include <kdbhelper.h>
+
+#include <ctype.h>
+#include <fnmatch.h>
+#include <stdlib.h>
+#include <string.h>
+
+static size_t strcnt (const char * s, char c)
+{
+	size_t count = 0;
+	while ((s = strchr (s + 1, c)) != NULL)
+	{
+		count++;
+	}
+	return count;
+}
+
+static char * elektraToFnmatchGlob (char * pattern)
+{
+	char * ptr = pattern;
+	while ((ptr = strchr (ptr, '/')) != NULL)
+	{
+		++ptr;
+		if ((*(ptr + 1) == '/' || *(ptr + 1) == '\0') && (*ptr == '#' || *ptr == '_'))
+		{
+			*ptr = '*'; // replace /#/ and /_/ with /*/
+		}
+	}
+	return pattern;
+}
+
+/**
+ * @brief checks whether the given string is a valid elektra array item name
+ *
+ * Valid array items consist of a '#' followed by <code>n</code> underscores ('_'),
+ * followed by <code>n+1</code> digits ('0'-'9'). Additionally the digits describe a
+ * valid 32-bit Integer (i.e. <code>0 <= x < 2^32</code>).
+ *
+ * @param name the string to check
+ * @retval true if @p name is a valid array item
+ * @retval false otherwise
+ */
+bool isArrayName (const char * name) // TODO: move? definitely useful elsewhere, e.g. reference plugin, libease, etc.
+{
+	if (name == NULL || *name != '#')
+	{
+		return false;
+	}
+
+	++name;
+
+	int underscores = 0;
+	while (*name == '_')
+	{
+		name++;
+		underscores++;
+	}
+
+	int digits = 0;
+	while (isdigit ((unsigned char) *name))
+	{
+		name++;
+		digits++;
+	}
+
+	return digits + underscores + 2 <= ELEKTRA_MAX_ARRAY_SIZE && digits == underscores + 1;
+}
+
+/**
+ * @brief checks whether a given Key matches a given globbing pattern
+ *
+ * The globbing patterns for this function are a superset of those from fnmatch(3)
+ * used with the FNM_PATHNAME flag:
+ * <ul>
+ * 	<li> '*' matches any series of characters other than '/'</li>
```

Comment:
Please either add character classes here or only describe the Elektra extensions.